src/Functions/array/arrayShiftRotate.cpp

```diff
@@ -0,0 +1,419 @@
+#include <Columns/ColumnArray.h>
+#include <DataTypes/DataTypeArray.h>
+#include <DataTypes/getLeastSupertype.h>
+#include <Functions/FunctionFactory.h>
+#include <Functions/FunctionHelpers.h>
+#include <Functions/IFunction.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int ILLEGAL_COLUMN;
+    extern const int ILLEGAL_TYPE_OF_ARGUMENT;
+    extern const int NUMBER_OF_ARGUMENTS_DOESNT_MATCH;
+}
+
+enum class ShiftRotateStrategy : uint8_t
+{
+    Shift,
+    Rotate
+};
+
+enum class ShiftRotateDirection : uint8_t
+{
+    Left,
+    Right
+};
+
+/// A helper structure for storing either a constant value or a full column and working with it.
+/// Useful for dealing with 'shift number' and 'default value' columns.
+struct ConstOrColumn
+{
+    ConstOrColumn() = default;
+    explicit ConstOrColumn(ColumnPtr src) { init(src); }
+
+    void init(ColumnPtr src)
+    {
+        if (isColumnConst(*src))
+            const_value = (*src)[0];
+        else
+            column = src;
+    }
+
+    Int64 getInt(size_t idx) const
```

Comment:
Doesn't `ColumnConst` without the wrapper provide desired behaviour? https://github.com/ClickHouse/ClickHouse/blob/5a4b001f8e8c8015a0df8686b2fad4055dcd2f1b/src/Columns/ColumnConst.h#L94-L97

Reply:
Do you mean the whole wrapper might be unneeded or just this function?

Reply:
Probably the whole wrapper. But if you had tried to go without wrapper and code was more complicated, then let's leave it. Although the initial purpose of `ColumnConst` was to provide `IColumn` interface for a single value.

Reply:
I see. Thanks! I used this pattern, because it was used in the other place (although without a wrapper).
I'll check if it can be dropped and will reply to this thread :)